crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md

```diff
@@ -0,0 +1,823 @@
+# error: "has no member `submodule`"
+reveal_type(mypackage.submodule.nested.X)  # revealed: Unknown
+```
+
+## Relative `from` Import of Direct Submodule in `__init__`, Mismatched Alias
+
+Renaming the submodule to something else disables the `__init__.pyi` idiom.
+
+`mypackage/__init__.pyi`:
+
+```pyi
+from . import imported as imported_m
+```
+
+`mypackage/imported.pyi`:
+
+```pyi
+X: int = 42
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+# error: "has no member `imported`"
+reveal_type(mypackage.imported.X)  # revealed: Unknown
+# error: "has no member `imported_m`"
+reveal_type(mypackage.imported_m.X)  # revealed: Unknown
+```
+
+## Relative `from` Import of Direct Submodule in `__init__`, Mismatched Alias (Non-Stub Check)
+
+`mypackage/__init__.py`:
+
+```py
+from . import imported as imported_m
+```
+
+`mypackage/imported.py`:
+
+```py
+X: int = 42
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+# TODO: this would be nice to support, as it works at runtime
+# error: "has no member `imported`"
+reveal_type(mypackage.imported.X)  # revealed: Unknown
+reveal_type(mypackage.imported_m.X)  # revealed: int
+```
+
+## Relative `from` Import of Direct Submodule in `__init__`, Matched Alias
+
+The `__init__.pyi` idiom should definitely always work if the submodule is renamed to itself, as
+this is the re-export idiom.
+
+`mypackage/__init__.pyi`:
+
+```pyi
+from . import imported as imported
+```
+
+`mypackage/imported.pyi`:
+
+```pyi
+X: int = 42
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+reveal_type(mypackage.imported.X)  # revealed: int
+```
+
+## Relative `from` Import of Direct Submodule in `__init__`, Matched Alias (Non-Stub Check)
+
+`mypackage/__init__.py`:
+
+```py
+from . import imported as imported
+```
+
+`mypackage/imported.py`:
+
+```py
+X: int = 42
+```
+
+`main.py`:
+
+```py
+import mypackage
+
+reveal_type(mypackage.imported.X)  # revealed: int
+```
+
+## Star Import Unaffected
+
+Even if the `__init__` idiom is in effect, star imports do not pick it up. (This is an arbitrary
+decision that mostly fell out of the implementation details and can be changed!)
+
+`mypackage/__init__.pyi`:
+
+```pyi
+from . import imported
+Z: int = 17
+```
+
+`mypackage/imported.pyi`:
+
+```pyi
+X: int = 42
+```
+
+`main.py`:
+
+```py
+from mypackage import *
+
+# TODO: this would be nice to support (available_submodule_attributes isn't visible to `*` imports)
+# error: "`imported` used when not defined"
+reveal_type(imported.X)  # revealed: Unknown
+reveal_type(Z)  # revealed: int
+```
+
+## Star Import Unaffected (Non-Stub Check)
+
+`mypackage/__init__.py`:
+
+```py
+from . import imported
+
+Z: int = 17
+```
+
+`mypackage/imported.py`:
+
+```py
+X: int = 42
+```
+
+`main.py`:
+
+```py
+from mypackage import *
+
+reveal_type(imported.X)  # revealed: int
+reveal_type(Z)  # revealed: int
+```
+
+## `from` Import of Non-Submodule
+
+A from import that terminates in a non-submodule should not expose the intermediate submodules as
+attributes.
```

Comment:
(this could also be changed in the future, as discussed, but no need to do that in this PR of course!)